rfcs/rdm-0045-versioning-support.md

```diff
@@ -0,0 +1,530 @@
+
+#### Updating ownership and permissions
+
+We propose adding a new method specifically update the permissions:
+
+```python
+service.update_permissions(id_, identity, data, links_config=...)
+service.update_owners(id_, identity, data, links_config=...)
+```
+
+- Resolve the id
+    - Should transparently handle the concept pid / pid issue (explicit - always use concept - not logical for APIs)
+- Check revision id
+- Check permissions
+- Load the data
+- Run components
+- Commit concept record
+- Index
+    - Reindex all records and drafts (bulk index)
+
+#### Search versions
+
+```python
+service.search_versions()
+```
+
+#### Reading a record/draft (dereferencing)
+
+(not needed? managed through record service)
+
+#### The service schema
+
+- A new schema for dumping permissions.
+
+### Presentation layer
+
+#### REST API
+
+The existing REST API, already define endpoints for the main operations discussed in the service:
+
+- ``GET /records`` - Search records.
+- ``GET /user/records`` - Search drafts.
+- ``POST /records`` - Create a new draft.
+- ``POST /records/:id/draft`` - Edit a record.
+- ``PUT /records/:id/draft`` - Update draft.
+- ``POST /records/:id/actions/publish`` - Publish a record.
+- ``GET /record/:id/versions`` - Search versions for a record.
+- ``POST /record/:id/versions`` - Create a new version.
+
+We will add a new endpoint for updating the permissions:
+
+- ``PUT /records/:id/access`` - Update permissions.
+
+We use ``PUT`` instead of e.g. ``PATCH /records/:id`` because we want to make use of the concept records version counter in the ETag for optimistic concurrency control, and we want to model the endpoint as a new REST resource. Some notes:
+
+- Identifier: We're using the ``:id`` of single version records instead of the concept record. This is to make it easier for API developers to deal only with a single identifier for a version, and to not confusing by using a different id.
+- Resource: We model an "access resource" instead of a "concept record resource" to make it easy to use and understand for API developers, and because different parts of the concept record may have different permissions as well as service methods.
+
+**Record/Draft Serialization**
+
+Both record and drafts serializations should include a link to the access resource if the user has manage permission:
+
+```json
+{
+    "id": "...",
+    "links": {
+        "access": "https://localhost:5000/api/records/:id/access",
+        "latest": "https://localhost:5000/api/records/:id/versions/latest/",
+        "latest_html": "https://localhost:5000/records/:id/latest/",
+    }
+}
+```
+
+**Access Serialization**
+
+The access serialization
+
+```json
+{
+    "id": "...",
+    "access": {
+        "owned_by": []
+        "grants": []
+    }
+    "links": {
+        "access": "https://localhost:5000/api/records/:id/access"
+    }
```

Comment:
this is itself right? Are we returning self-referential links for the files metadata endpoint? This is the endpoint that is the most similar. For records we call these links `self`, but is this really needed in this case...?